server/src/main/java/org/elasticsearch/index/mapper/NestedValueFetcher.java

```diff
@@ -0,0 +1,90 @@
+import java.util.Map;
+import java.util.Set;
+
+public class NestedValueFetcher implements ValueFetcher {
+
+    private final String nestedFieldPath;
+    private final FieldFetcher nestedFieldFetcher;
+
+    // the name of the nested field without the full path, i.e. in foo.bar.baz it would be baz
+    private final String nestedFieldName;
+    private final String[] nestedPathParts;
+
+    public NestedValueFetcher(String nestedField, FieldFetcher nestedFieldFetcher) {
+        assert nestedField != null && nestedField.isEmpty() == false;
+        this.nestedFieldPath = nestedField;
+        this.nestedFieldFetcher = nestedFieldFetcher;
+        this.nestedPathParts = nestedFieldPath.split("\\.");
+        this.nestedFieldName = nestedPathParts[nestedPathParts.length - 1];
+    }
+
+    @Override
+    public List<Object> fetchValues(SourceLookup lookup, Set<String> ignoreFields) throws IOException {
+        List<Object> nestedEntriesToReturn = new ArrayList<>();
+        Map<String, Object> filteredSource = new HashMap<>();
+        Map<String, Object> stub = createSourceMapStub(filteredSource);
+        List<?> nestedValues = XContentMapValues.extractNestedValue(nestedFieldPath, (lookup.source()));
+        if (nestedValues == null) {
+            return Collections.emptyList();
+        }
+        for (Object entry : nestedValues) {
+            // add this one entry only to the stub and use this as source lookup
+            stub.put(nestedFieldName, entry);
+            SourceLookup nestedSourceLookup = new SourceLookup();
+            nestedSourceLookup.setSource(filteredSource);
+
+            Map<String, DocumentField> fetchResult = nestedFieldFetcher.fetch(nestedSourceLookup, ignoreFields);
+
+            Map<String, Object> nestedEntry = new HashMap<>();
+            for (DocumentField field : fetchResult.values()) {
+                List<Object> fetchValues = field.getValues();
+                if (fetchValues.isEmpty() == false) {
+                    String keyInNestedMap = field.getName().substring(nestedFieldPath.length() + 1);
+                    nestedEntry.put(keyInNestedMap, fetchValues);
+                }
+            }
+            if (nestedEntry.isEmpty() == false) {
+                nestedEntriesToReturn.add(nestedEntry);
+            }
+        }
+        return nestedEntriesToReturn;
+    }
+
+    // create a filtered source map stub which contains the nested field path
+    private Map<String, Object> createSourceMapStub(Map<String, Object> filteredSource) {
+        Map<String, Object> next = filteredSource;
+        for (int i = 0; i < nestedPathParts.length - 1; i++) {
+            String part = nestedPathParts[i];
+            Map<String, Object> newMap = new HashMap<>();
+            next.put(part, newMap);
+            next = newMap;
+        }
+        return next;
+    }
+
+    @Override
+    public void setNextReader(LeafReaderContext context) {
+        this.nestedFieldFetcher.setNextReader(context);
```

Comment:
I realized that DocValueFetcher implements "setNextReader" and probably needs to be called, even if they are nested inside the FieldFetcher inside this NestedValueFetcher, so we probably need to pass it on here. Haven't found a reliable way to test what happens if we don't though...

Reply:
I indeed tried removing this and didn't see any tests fail. Since this PR is already complex, what if we backed out the changes related to `DocValueFetcher` and tackled this in an immediate follow-up? I think the changes include this, plus making the nested `SourceLookup` return `docId`.

Reply:
This problem surfaced when running the yaml tests as part of the runtime-fields qa tests, which are now muted because we suspect a bug in their test infra. However the NPE I got before was happening if we somehow have a DocValueFetcher inside a nested field fetcher. One way to demonstrate this would be to add a token_count field under a nested field in the yaml tests. Currently this would return no values (another bug unrelated to this PR) but without this would raise an NPE. I'm happy to take this out to another follow up PR but wouldn't like to drop this.

Reply:
+1 to tackle this in an immediate follow-up! I agree it's serious and we should make sure not to drop it, but also rare (using `token_count` in nested).